File: architecture/modules/storage/node/src/extensions.ts
import {StorageError,call,type StorageProvider} from './contracts.ts';
// Server-side extension port. A project adapter must implement and explicitly advertise its operation.
export function runStorageExtension<T=unknown>(provider:StorageProvider,name:string,input:unknown):Promise<T>{
 if(!/^[a-z][a-z0-9.-]{1,63}$/.test(name)||Buffer.byteLength(JSON.stringify(input)??'')>65536)throw new StorageError('INVALID_INPUT');
 if(!provider.capabilities.extensions.includes(name)||!provider.extension)throw new StorageError('UNSUPPORTED');
 return call(()=>provider.extension!<T>(name,input));
}

File: architecture/modules/storage/go/extensions.go
package storage

import (
	"context"
	"encoding/json"
	"regexp"
)

type ExtensionProvider interface {
	Extension(context.Context, string, json.RawMessage) (json.RawMessage, error)
}

func RunStorageExtension(ctx context.Context, provider Provider, name string, input json.RawMessage) (json.RawMessage, error) {
	if !regexp.MustCompile("^[a-z][a-z0-9.-]{1,63}$").MatchString(name) || len(input) > 65536 || !json.Valid(input) {
		return nil, fail("INVALID_INPUT")
	}
	supported := false
	for _, operation := range provider.Capabilities().Extensions {
		if operation == name {
			supported = true
		}
	}
	adapter, ok := provider.(ExtensionProvider)
	if !supported || !ok {
		return nil, fail("UNSUPPORTED")
	}
	output, e := adapter.Extension(ctx, name, input)
	return output, Normalize(e)
}
